File: src/main/java/work/lclpnet/ap2/impl/game/data/Ordering.java
package work.lclpnet.ap2.impl.game.data;

import java.util.Comparator;
import java.util.OptionalInt;
import java.util.function.ToIntFunction;
import java.util.stream.IntStream;

public enum Ordering {

    DESCENDING,
    ASCENDING;

    public OptionalInt best(IntStream stream) {
        return switch (this) {
            case DESCENDING -> stream.max();
            case ASCENDING -> stream.min();
        };
    }

    public Ordering opposite() {
        return switch (this) {
            case DESCENDING -> ASCENDING;
            case ASCENDING -> DESCENDING;
        };
    }

    public <T> Comparator<T> order(ToIntFunction<T> keyExtractor) {
        var comparator = Comparator.comparingInt(keyExtractor);

        return switch (this) {
            case DESCENDING -> comparator.reversed();
            case ASCENDING -> comparator;
        };
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/data/Ordering.kt
package work.lclpnet.ap2.game.data

import java.util.function.ToDoubleFunction
import java.util.function.ToIntFunction
import java.util.stream.DoubleStream
import java.util.stream.IntStream

enum class Ordering {
    DESCENDING,
    ASCENDING;

    fun best(stream: IntStream): Int? {
        return (when (this) {
            DESCENDING -> stream.max()
            ASCENDING -> stream.min()
        }).stream().boxed().findAny().orElse(null)
    }

    fun best(stream: DoubleStream): Double? {
        return (when (this) {
            DESCENDING -> stream.max()
            ASCENDING -> stream.min()
        }).stream().boxed().findAny().orElse(null)
    }

    fun opposite(): Ordering {
        return when (this) {
            DESCENDING -> ASCENDING
            ASCENDING -> DESCENDING
        }
    }

    fun <T> orderInt(keyExtractor: ToIntFunction<T>): Comparator<T> {
        val comparator = Comparator.comparingInt(keyExtractor)

        return when (this) {
            DESCENDING -> comparator.reversed()
            ASCENDING -> comparator
        }
    }

    fun <T> orderDouble(keyExtractor: ToDoubleFunction<T>): Comparator<T> {
        val comparator = Comparator.comparingDouble(keyExtractor)

        return when (this) {
            DESCENDING -> comparator.reversed()
            ASCENDING -> comparator
        }
    }
}
